src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -727,11 +744,159 @@ BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementT
+            string? methodName,
+            BindingDiagnosticBag diagnostics)
+        {
+            if (!(builderType is NamedTypeSymbol { TypeKind: TypeKind.Class or TypeKind.Struct, IsGenericType: false }))
+            {
+                return null;
+            }
+
+            if (string.IsNullOrEmpty(methodName))
+            {
+                return null;
+            }
+
+            foreach (var candidate in builderType.GetMembers(methodName))
+            {
+                var method = candidate as MethodSymbol;
+                if (method is null || !method.IsStatic)
+                {
+                    continue;
+                }
+
+                var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                if (!IsAccessible(method, ref useSiteInfo))
+                {
+                    continue;
+                }
+
+                var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance();
+                targetType.GetAllTypeArgumentsNoUseSiteDiagnostics(builder);
+                var allTypeArguments = builder.ToImmutableAndFree();
+
+                if (method.Arity != allTypeArguments.Length)
+                {
+                    continue;
+                }
+
+                if (!(method.Parameters is [var parameter] && isReadOnlySpanType(Compilation, parameter.Type)))
+                {
+                    continue;
+                }
+
+                if (allTypeArguments.Length > 0)
+                {
+                    var allTypeParameters = TypeMap.TypeParametersAsTypeSymbolsWithAnnotations(targetType.OriginalDefinition.GetAllTypeParameters());
+                    var mDef = method.OriginalDefinition.Construct(allTypeParameters);
+                    var spanElementType = ((NamedTypeSymbol)mDef.Parameters[0].Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+                    if (!elementTypeOriginalDefinition.Equals(spanElementType, TypeCompareKind.AllIgnoreOptions))
+                    {
+                        continue;
+                    }
+                    if (!targetType.OriginalDefinition.Equals(mDef.ReturnType, TypeCompareKind.AllIgnoreOptions))
+                    {
+                        continue;
+                    }
+                    method = method.Construct(allTypeArguments);
+                }
+
+                if (!targetType.Equals(method.ReturnType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    continue;
+                }
+
+                diagnostics.Add(syntax, useSiteInfo);
+                return method;
+            }
+
+            return null;
+
+            static bool isReadOnlySpanType(CSharpCompilation compilation, TypeSymbol type)
+            {
+                return type is NamedTypeSymbol { Arity: 1 } namedType
+                    && namedType.OriginalDefinition.Equals(compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.AllIgnoreOptions);
```

Comment:
We had to get `ROS` from the compilation earlier for an assert, and now we're doing it potentially a number of times here. Consider just extracting this and doing it up front.